src/corelibs/U2View/src/ov_msa/exclude_list/MsaExcludeList.cpp

```diff
@@ -0,0 +1,593 @@
+        insertionIndex = msaEditor->getSelection().getRectList().last().bottom() + 1;
+    }
+    AddSequenceObjectsToAlignmentTask task(msaObject, sequences, insertionIndex, true);
+    task.run();  // TODO: extract as an utility method.
+    if (!task.hasError()) {
+        trackedUndoMsaVersions.insert(versionBefore, {false, excludeListRowIdsMovedToMsa});
+        trackedRedoMsaVersions.insert(msaObject->getObjectVersion(), {false, excludeListRowIdsMovedToMsa});
+    }
+    updateState();
+}
+
+void MsaExcludeListWidget::handleUndoRedoInMsaEditor(const MultipleAlignment &maBefore, const MaModificationInfo &modInfo) {
+    auto msaObject = msaEditor->getMaObject();
+    int version = msaObject->getObjectVersion();
+    if (modInfo.type != MaModificationType_Undo && modInfo.type != MaModificationType_Redo) {
+        auto truncateVersionData = [version](QMap<int, UndoRedoStep> &undoRedoMap) {
+            CHECK(!undoRedoMap.isEmpty() && version <= undoRedoMap.lastKey(), );
+            QList<int> allUndoRedoVersions = undoRedoMap.keys();
+            for (int undoRedoVersion : qAsConst(allUndoRedoVersions)) {
+                if (undoRedoVersion >= version) {
+                    undoRedoMap.remove(undoRedoVersion);
+                }
+            }
+        };
+        truncateVersionData(trackedUndoMsaVersions);
+        truncateVersionData(trackedRedoMsaVersions);
+        return;
+    }
+    bool isRedo = modInfo.type == MaModificationType_Redo;
+    CHECK((isRedo && trackedRedoMsaVersions.contains(version)) || (!isRedo && trackedUndoMsaVersions.contains(version)), );
+
+    const UndoRedoStep &undoRedoContext = isRedo ? trackedRedoMsaVersions.value(version) : trackedUndoMsaVersions.value(version);
+    bool isAddToExcludeList = (isRedo && undoRedoContext.isMoveFromMsaToExcludeList) || (!isRedo && !undoRedoContext.isMoveFromMsaToExcludeList);
+    if (isAddToExcludeList) {  // Add rows removed from MSA to Exclude list
+        QList<MultipleAlignmentRow> msaRows;
+        QSet<qint64> msaRowIdsAfter = msaObject->getRowIds().toSet();
+        for (int i = 0; i < maBefore->getNumRows(); i++) {
+            const MultipleAlignmentRow &row = maBefore->getRow(i);
+            if (!msaRowIdsAfter.contains(row->getRowId())) {
+                msaRows << row;
+            }
+        }
+        SAFE_POINT(msaRows.size() == undoRedoContext.excludeListRowIdsDelta.size(), "Failed to map Exclude List rows to Msa rows", )
+        for (int i = 0; i < msaRows.size(); i++) {
+            auto &msaRow = msaRows[i];
+            int excludeListRowId = undoRedoContext.excludeListRowIdsDelta[i];
+            addMsaRowEntry(msaRow, excludeListRowId);
+        }
+    } else {  // Remove rows from Exclude List.
+        QSet<int> rowIdsToRemove = undoRedoContext.excludeListRowIdsDelta.toSet();
+        QList<QListWidgetItem *> listItemsToRemove;
+        for (int rowIndex = nameListView->count(); --rowIndex >= 0;) {
```

Comment:
I suppose, it would be more clear to have a for like this:
`for (int rowIndex = nameListView->count() - 1; rowIndex >= 0; rowIndex--) {`
Otherwise it's some kind of yoda-style, which you do not like so much.

Reply:
A simple reg-ex in UGENE code shows >100 backward loops like this. This notion is equal to and  shorten than what you suggested, so I would keep it. 
Compare:
```for (int rowIndex = nameListView->count(); --rowIndex >= 0; ) ```

```for (int rowIndex = nameListView->count() - 1; rowIndex >= 0; rowIndex--) ```

However I do not mind if you use the other variant.